cluster-master.js

```diff
@@ -306,20 +320,56 @@ function forkListener () {
   })
 }
 
+function handleWorkerCondemnedToBeDead(workers) {
+  if (!restarting){
+    debug("handler of worker condemned to be dead is started")
+      Object.keys(workers).forEach( function(key) {
+        if (workers[key].willBeDead == true) {
+          debug("force disconnect worker id:", key)
+        workers[key].process.disconnect()
+        }
+        // delaye the force disconnect for next time for suicide case
+        if (workers[key].suicide == true) {
+          debug("set willBeDead in worker id:", key)
+          workers[key].willBeDead = true
+        }
+      });
+  }
+}
+
+setInterval(function() {
+  handleWorkerCondemnedToBeDead(cluster.workers)
+}, handleWorkerCondemnedToBeDeadInterval);
+
+
 function restart (cb) {
-  if (restarting) {
-    debug("Already restarting.  Cannot restart yet.")
+
+  if (restarting || tooQuick) {
+    debug("Already restarting or too quick restart.  Cannot restart yet.")
     return
   }
 
+  // cleanUp before restarting
+  handleWorkerCondemnedToBeDead(cluster.workers)
+
   restarting = true
+  // prevent too quick reload (30s)
+  tooQuick = true
+  setTimeout(function() {
+    tooQuick = false
+  }, tooQuickTimeOut)
 
   // graceful restart.
   // all the existing workers get killed, and this
   // causes new ones to be spawned.  If there aren't
   // already the intended number, then fork new extras.
-  var current = Object.keys(cluster.workers)
-  , length = current.length
+  // Apply restart only on worker that are not tagged with willBeDead
+  // this will prevent the growth in size when restart is fired
+  var current = _.filter(Object.keys(cluster.workers), function(workerId){
```

Comment:
the `current` is only the list of `ID` of the workers, so i just need to do a `filter` on it